### PR TITLE
Implementation of UUID

### PR DESCRIPTION
**Pull request by Thumimku**
## Purpose
Current implementation using the auto-increment id for both remote fetch configuration and deployment revision. The purpose of this PR is to change the auto-increment id to UUID, not only in database tables, but also in java level. 

## Related PRs
https://github.com/wso2/carbon-identity-framework/pull/2891
## Automation tests
 - Unit tests 
   >passed all unit tests

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

- H2
- MSSQL
- MySQL
- Oracle 12c
- DB2
- Postgres

### CODE REVIEW

File: components/org.wso2.carbon.identity.remotefetch.core.ui/src/main/resources/web/remotefetch-core/delete-remotefetch-config.jsp [hunk 1]
```diff
@@ -22,22 +22,14 @@
 <%
     
     if (request.getParameter("id") != null) {
-        int id = -1;
-        
+        String id = request.getParameter("id");
```
Comment (piraveena) [inline components/org.wso2.carbon.identity.remotefetch.core.ui/src/main/resources/web/remotefetch-core/delete-remotefetch-config.jsp:25]: Better to move this "id" to a constant

File: components/org.wso2.carbon.identity.remotefetch.core/src/main/java/org/wso2/carbon/identity/remotefetch/core/RemoteFetchConfigurationServiceImpl.java [hunk 2]
```diff
@@ -52,8 +53,13 @@ public ValidationReport addRemoteFetchConfiguration(RemoteFetchConfiguration fet
                         .getRemoteFetchComponentRegistry(), fetchConfiguration);
 
         ValidationReport validationReport = validator.validate();
+//        if (log.isDebugEnabled()) {
```
Comment (piraveena) [inline components/org.wso2.carbon.identity.remotefetch.core/src/main/java/org/wso2/carbon/identity/remotefetch/core/RemoteFetchConfigurationServiceImpl.java:56]: Add a debug log if it is required. Or remove this commented block

File: components/org.wso2.carbon.identity.remotefetch.core.ui/src/main/resources/web/remotefetch-core/delete-remotefetch-config.jsp (hunk 1 above)
Comment (Thumimku) [reply components/org.wso2.carbon.identity.remotefetch.core.ui/src/main/resources/web/remotefetch-core/delete-remotefetch-config.jsp:25]: This one time used a local parameter, Can you please explin we need to move as constant in this JSP file?

File: components/org.wso2.carbon.identity.remotefetch.common/src/main/java/org/wso2/carbon/identity/remotefetch/common/ValidationReport.java [hunk 3]
```diff
@@ -131,6 +133,22 @@ public void setValidationStatus(ValidationStatus validationStatus) {
         this.validationStatus = validationStatus;
     }
 
+    /**
+     * get id
+     * @return id
+     */
+    public String getId() {
```
[the remark this answers is not in the document]
Comment (Thumimku) [reply components/org.wso2.carbon.identity.remotefetch.common/src/main/java/org/wso2/carbon/identity/remotefetch/common/ValidationReport.java:140]: Resolved

[the remark this answers is not in the document]
Comment (Thumimku) [reply components/org.wso2.carbon.identity.remotefetch.common/src/main/java/org/wso2/carbon/identity/remotefetch/common/ValidationReport.java:140]: Resolved

File: features/org.wso2.carbon.identity.remotefetch.core.feature/resources/dbscripts/oracle.sql [hunk 4]
```diff
@@ -33,16 +22,4 @@ CREATE TABLE IDN_REMOTE_FETCH_REVISIONS(
...
-  BEFORE INSERT
-  ON IDN_REMOTE_FETCH_REVISIONS
-  REFERENCING NEW AS NEW
-  FOR EACH ROW
-  BEGIN
-    SELECT IDN_REMOTE_FETCH_REVISIONS_SEQ.nextval INTO :NEW.ID FROM dual;
-  END;
-/
\ No newline at end of file
+)/
```
Comment (piraveena) [inline features/org.wso2.carbon.identity.remotefetch.core.feature/resources/dbscripts/oracle.sql:25]: add a new line here

Comment (Thumimku) [reply features/org.wso2.carbon.identity.remotefetch.core.feature/resources/dbscripts/oracle.sql]: Resolved

File: components/org.wso2.carbon.identity.remotefetch.core.ui/src/main/resources/web/remotefetch-core/add-remotefetch-config.jsp [hunk 5]
```diff
@@ -35,20 +35,14 @@
     pageContext.setAttribute("fetchConfigurationJs", null);
     
     if (request.getParameter("id") != null) {
```
Comment (piraveena) [inline components/org.wso2.carbon.identity.remotefetch.core.ui/src/main/resources/web/remotefetch-core/add-remotefetch-config.jsp:37]: Make it as String editId = request.getParameter("id"); and then check for null. if (editId!=null).

Calling this  request.getParameter("id") two times is redundant

File: components/org.wso2.carbon.identity.remotefetch.core.ui/src/main/resources/web/remotefetch-core/delete-remotefetch-config.jsp (hunk 1 above)
Comment (Thumimku) [reply components/org.wso2.carbon.identity.remotefetch.core.ui/src/main/resources/web/remotefetch-core/delete-remotefetch-config.jsp]: Resolved

File: components/org.wso2.carbon.identity.remotefetch.core.ui/src/main/resources/web/remotefetch-core/add-remotefetch-config.jsp (hunk 5 above)
Comment (Thumimku) [reply components/org.wso2.carbon.identity.remotefetch.core.ui/src/main/resources/web/remotefetch-core/add-remotefetch-config.jsp]: Resolved

File: components/org.wso2.carbon.identity.remotefetch.common/src/main/java/org/wso2/carbon/identity/remotefetch/common/BasicRemoteFetchConfiguration.java [hunk 6]
```diff
@@ -53,16 +53,16 @@ public BasicRemoteFetchConfiguration(int id, boolean isEnabled,
...
      */
-    public int getId() {
+    public String getId() {
 
         return id;
     }
 
     /**
-     *get remoteFetchName.
+     *Get remoteFetchName.
```
Comment (jeradrutnam) [inline components/org.wso2.carbon.identity.remotefetch.common/src/main/java/org/wso2/carbon/identity/remotefetch/common/BasicRemoteFetchConfiguration.java:65]: Indentation issue

File: components/org.wso2.carbon.identity.remotefetch.common/src/main/java/org/wso2/carbon/identity/remotefetch/common/ValidationReport.java [hunk 7]
```diff
@@ -28,13 +28,15 @@
  */
 public class ValidationReport {
 
+
```
Comment (jeradrutnam) [inline components/org.wso2.carbon.identity.remotefetch.common/src/main/java/org/wso2/carbon/identity/remotefetch/common/ValidationReport.java:31]: Unwanted new line

File: components/org.wso2.carbon.identity.remotefetch.core.ui/src/main/resources/web/remotefetch-core/add-remotefetch-config-insert.jsp [hunk 8]
```diff
@@ -42,7 +43,7 @@
                 validationReport = RemoteFetchConfigurationClient.addFetchConfiguration(payload, currentUser);
                 CarbonUIMessage.sendCarbonUIMessage("Configuration successfully added!", CarbonUIMessage.INFO, request);
             } else if (action.equalsIgnoreCase("update")) {
-                validationReport = RemoteFetchConfigurationClient.updateFetchConfiguration(payload, currentUser);
+                validationReport = RemoteFetchConfigurationClient.updateFetchConfiguration(id,payload, currentUser);
```
Comment (jeradrutnam) [inline components/org.wso2.carbon.identity.remotefetch.core.ui/src/main/resources/web/remotefetch-core/add-remotefetch-config-insert.jsp:46]: Keep a space after `updateFetchConfiguration(id,`

File: components/org.wso2.carbon.identity.remotefetch.core/src/test/java/org/wso2/carbon/identity/remotefetch/core/dao/impl/DeploymentRevisionDAOImplTest.java [hunk 9]
```diff
@@ -45,17 +45,20 @@
...
     private static final String DB_NAME = "IDN_REMOTE_FETCH_DB";
-    private int configId = 1;
+    private String configId = "00000000-0000-0000-0000-d29bed62f7bd";
+    private String  deploymentRevisionId = "11111111-0000-0000-0000-d29bed62f7bd";
 
     DeploymentRevisionDAOImpl deploymentRevisionDAO = new DeploymentRevisionDAOImpl();
     DeploymentRevision deploymentRevision = new DeploymentRevision(configId, null);
 
+
+
```
Comment (NipuniBhagya) [inline components/org.wso2.carbon.identity.remotefetch.core/src/test/java/org/wso2/carbon/identity/remotefetch/core/dao/impl/DeploymentRevisionDAOImplTest.java:61]: Unwanted new line.

File: components/org.wso2.carbon.identity.remotefetch.core/src/test/java/org/wso2/carbon/identity/remotefetch/core/dao/impl/RemoteFetchConfigurationDAOImplTest.java [hunk 10]
```diff
@@ -42,13 +42,13 @@
...
 /**
- * Unit test covering RemoteFetchConfigurationDAOImpl
+ * Unit test covering RemoteFetchConfigurationDAOImpl.
  */
 @PrepareForTest(JdbcUtils.class)
 public class RemoteFetchConfigurationDAOImplTest extends PowerMockTestCase {
 
     private static final String DB_NAME = "IDN_REMOTE_FETCH_DB";
-    private int remoteFetchConfigurationId = 1;
+    private String remoteFetchConfigurationId = "00000000-0000-0000-0000-d29bed62f7bd";
```
Comment (NipuniBhagya) [inline components/org.wso2.carbon.identity.remotefetch.core/src/test/java/org/wso2/carbon/identity/remotefetch/core/dao/impl/RemoteFetchConfigurationDAOImplTest.java:51]: Shouldn't we assign this to constant and use it here?

File: features/org.wso2.carbon.identity.remotefetch.core.feature/resources/dbscripts/db2.sql [hunk 11]
```diff
@@ -1,7 +1,7 @@
 CREATE TABLE IDN_REMOTE_FETCH_CONFIG(
-    ID INTEGER NOT NULL,
+	ID VARCHAR(255) NOT NULL,
```
Comment (NipuniBhagya) [inline features/org.wso2.carbon.identity.remotefetch.core.feature/resources/dbscripts/db2.sql:2]: Indentation issue.

File: features/org.wso2.carbon.identity.remotefetch.core.feature/resources/dbscripts/db2.sql [hunk 12]
```diff
@@ -36,14 +25,3 @@ CREATE TABLE IDN_REMOTE_FETCH_REVISIONS (
...
- /
-CREATE TRIGGER IDN_REMOTE_FETCH_REVISIONS_TRIG NO CASCADE
-BEFORE INSERT
-ON IDN_REMOTE_FETCH_REVISIONS
-REFERENCING NEW AS NEW
-FOR EACH ROW MODE DB2SQL
-  BEGIN ATOMIC
-SET (NEW.ID) = (NEXTVAL FOR IDN_REMOTE_FETCH_REVISIONS_SEQ);
-END
-/
```
Comment (NipuniBhagya) [inline features/org.wso2.carbon.identity.remotefetch.core.feature/resources/dbscripts/db2.sql:49]: Add a new line here.

File: components/org.wso2.carbon.identity.remotefetch.core/src/main/java/org/wso2/carbon/identity/remotefetch/core/dao/impl/RemoteFetchConfigurationDAOImpl.java [hunk 13]
```diff
@@ -44,7 +45,7 @@
 import static org.wso2.carbon.identity.remotefetch.core.RemoteFetchConstants.FACTOR_INDENT;
 
 /**
- * this class accesses IDN_REMOTE_FETCH_CONFIG table to store/update and delete Remote Fetch configurations.
+ * This class accesses IDN_REMOTE_FETCH_CONFIG table to store/update and delete Remote Fetch configurations.
  * TODO : Implement Name preparedstatement
```
Comment (jeradrutnam) [inline components/org.wso2.carbon.identity.remotefetch.core/src/main/java/org/wso2/carbon/identity/remotefetch/core/dao/impl/RemoteFetchConfigurationDAOImpl.java:49]: Shall we implement this TODO ?

File: components/org.wso2.carbon.identity.remotefetch.core/src/main/java/org/wso2/carbon/identity/remotefetch/core/dao/impl/RemoteFetchConfigurationDAOImpl.java [hunk 14]
```diff
@@ -55,17 +56,31 @@
...
+                            preparedStatement -> {
+                                preparedStatement.setString(1, configuration.getRemoteFetchConfigurationId());
+                                preparedStatement.setInt(2, configuration.getTenantId());
+                                preparedStatement.setString(3, (configuration.isEnabled() ? "1" : "0"));
+                                preparedStatement.setString(4, configuration.getUserName());
+                                preparedStatement.setString(5, configuration.getRepositoryManagerType());
+                                preparedStatement.setString(6, configuration.getActionListenerType());
+                                preparedStatement.setString(7, configuration.getConfigurationDeployerType());
+
+                                //Encode object attributes to JSON
```
Comment (jeradrutnam) [inline components/org.wso2.carbon.identity.remotefetch.core/src/main/java/org/wso2/carbon/identity/remotefetch/core/dao/impl/RemoteFetchConfigurationDAOImpl.java:75]: `// Encode object attributes to JSON`

File: components/org.wso2.carbon.identity.remotefetch.core/src/main/java/org/wso2/carbon/identity/remotefetch/core/dao/impl/RemoteFetchConfigurationDAOImpl.java [hunk 15]
```diff
@@ -55,17 +56,31 @@
...
+                                JSONObject attributesBundle = this.makeAttributeBundle(configuration);
+
+                                preparedStatement.setString(8, attributesBundle.toString(FACTOR_INDENT));
+                                preparedStatement.setString(9, configuration.getRemoteFetchName());
+                                }
+                            , configuration, false)
+            ;
```
Comment (jeradrutnam) [inline components/org.wso2.carbon.identity.remotefetch.core/src/main/java/org/wso2/carbon/identity/remotefetch/core/dao/impl/RemoteFetchConfigurationDAOImpl.java:82]:
```
                                }, configuration, false);
```

File: components/org.wso2.carbon.identity.remotefetch.core/src/test/java/org/wso2/carbon/identity/remotefetch/core/dao/TestConstants.java [hunk 16]
```diff
@@ -38,7 +38,7 @@
 
     public static JSONObject getAttributesJson() {
 
-        int remoteFetchConfigurationId = -1;
+        String remoteFetchConfigurationId = "00000000-0000-0000-0000-d29bed62f7bd";
```
Comment (jeradrutnam) [inline components/org.wso2.carbon.identity.remotefetch.core/src/test/java/org/wso2/carbon/identity/remotefetch/core/dao/TestConstants.java:41]: Use a constant here

File: components/org.wso2.carbon.identity.remotefetch.core/src/test/java/org/wso2/carbon/identity/remotefetch/core/impl/deployers/config/ServiceProviderConfigDeployerTest.java [hunk 17]
```diff
@@ -54,7 +54,7 @@
 
     private static final int TENANT_ID = -1234;
     private static final String USER_NAME = "admin";
-    private static final int ID = 1;
+    private static final String ID = "00000000-0000-0000-0000-d29bed62f7bd";
```
Comment (jeradrutnam) [inline components/org.wso2.carbon.identity.remotefetch.core/src/test/java/org/wso2/carbon/identity/remotefetch/core/impl/deployers/config/ServiceProviderConfigDeployerTest.java:57]: Use a constant here

File: components/org.wso2.carbon.identity.remotefetch.core/src/test/java/org/wso2/carbon/identity/remotefetch/core/impl/handlers/action/PollingActionListenerBuilderTest.java [hunk 18]
```diff
@@ -69,7 +69,8 @@
     ConfigDeployer configDeployer;
 
     Map<String, String> actionListenerAttributesMap = new HashMap<>();
-    RemoteFetchConfiguration remoteFetchConfiguration = new RemoteFetchConfiguration(1,
+    RemoteFetchConfiguration remoteFetchConfiguration = new RemoteFetchConfiguration
+            ("00000000-0000-0000-0000-d29bed62f7bd",
```
Comment (jeradrutnam) [inline components/org.wso2.carbon.identity.remotefetch.core/src/test/java/org/wso2/carbon/identity/remotefetch/core/impl/handlers/action/PollingActionListenerBuilderTest.java:73]: Add it to a constant and use

File: components/org.wso2.carbon.identity.remotefetch.core/src/test/java/org/wso2/carbon/identity/remotefetch/core/impl/handlers/action/PollingActionListenerBuilderTest.java [hunk 19]
```diff
@@ -136,11 +137,12 @@ public void testBuildForException() throws Exception {
 
     private DeploymentRevision buildDummyDeploymentRevision() {
 
-        DeploymentRevision deploymentRevision = new DeploymentRevision(1, null);
+        DeploymentRevision deploymentRevision = new DeploymentRevision
+                ("00000000-0000-0000-0000-d29bed62f7bd", null);
```
Comment (jeradrutnam) [inline components/org.wso2.carbon.identity.remotefetch.core/src/test/java/org/wso2/carbon/identity/remotefetch/core/impl/handlers/action/PollingActionListenerBuilderTest.java:141]: Add it to a constant and use

File: components/org.wso2.carbon.identity.remotefetch.core/src/test/java/org/wso2/carbon/identity/remotefetch/core/impl/handlers/action/PollingActionListenerBuilderTest.java [hunk 20]
```diff
@@ -136,11 +137,12 @@ public void testBuildForException() throws Exception {
...
         long millis = System.currentTimeMillis();
         java.sql.Date date = new java.sql.Date(millis);
 
-        deploymentRevision.setDeploymentRevisionId(1);
+        deploymentRevision.setDeploymentRevisionId("11111111-0000-0000-0000-d29bed62f7bd");
```
Comment (jeradrutnam) [inline components/org.wso2.carbon.identity.remotefetch.core/src/test/java/org/wso2/carbon/identity/remotefetch/core/impl/handlers/action/PollingActionListenerBuilderTest.java:145]: Please check this on all the files and replace with a constant

File: components/org.wso2.carbon.identity.remotefetch.core/src/test/java/org/wso2/carbon/identity/remotefetch/core/util/RemoteFetchConfigurationValidatorTest.java [hunk 21]
```diff
@@ -32,12 +32,12 @@
...
 
 /**
- * Unit test covering RemoteFetchConfigurationValidator
+ * Unit test covering RemoteFetchConfigurationValidator.
  */
 @PrepareForTest(RemoteFetchConfigurationValidator.class)
 public class RemoteFetchConfigurationValidatorTest extends PowerMockTestCase {
 
-    private static final int ID = 1;
+    private static final String ID = "00000000-0000-0000-0000-d29bed62f7bd";
```
Comment (jeradrutnam) [inline components/org.wso2.carbon.identity.remotefetch.core/src/test/java/org/wso2/carbon/identity/remotefetch/core/util/RemoteFetchConfigurationValidatorTest.java:40]: Same here

File: features/org.wso2.carbon.identity.remotefetch.core.feature/resources/dbscripts/mysql.sql [hunk 22]
```diff
@@ -12,8 +12,8 @@ CREATE TABLE IDN_REMOTE_FETCH_CONFIG (
 )ENGINE INNODB;
 
 CREATE TABLE IDN_REMOTE_FETCH_REVISIONS (
-	ID INT NOT NULL AUTO_INCREMENT,
-	CONFIG_ID INT NOT NULL,
+    ID VARCHAR(255) NOT NULL,
+    CONFIG_ID VARCHAR(255) NOT NULL,
```
Comment (jeradrutnam) [inline features/org.wso2.carbon.identity.remotefetch.core.feature/resources/dbscripts/mysql.sql:16]: Indentation issue

**Pull request merged**